tcl8.3.4/doc/DString.3
'\"
'\" Copyright (c) 1993 The Regents of the University of California.
'\" Copyright (c) 1994-1996 Sun Microsystems, Inc.
'\"
'\" See the file "license.terms" for information on usage and redistribution
'\" of this file, and for a DISCLAIMER OF ALL WARRANTIES.
'\" 
'\" RCS: @(#) $Id: DString.3,v 1.1 2001/11/05 20:06:32 jgdavidson Exp $
'\" 
.so man.macros
.TH Tcl_DString 3 7.4 Tcl "Tcl Library Procedures"
.BS
.SH NAME
Tcl_DStringInit, Tcl_DStringAppend, Tcl_DStringAppendElement, Tcl_DStringStartSublist, Tcl_DStringEndSublist, Tcl_DStringLength, Tcl_DStringValue, Tcl_DStringSetLength, Tcl_DStringFree, Tcl_DStringResult, Tcl_DStringGetResult \- manipulate dynamic strings
.SH SYNOPSIS
.nf
\fB#include <tcl.h>\fR
.sp
\fBTcl_DStringInit\fR(\fIdsPtr\fR)
.sp
char *
\fBTcl_DStringAppend\fR(\fIdsPtr, string, length\fR)
.sp
char *
\fBTcl_DStringAppendElement\fR(\fIdsPtr, string\fR)
.sp
\fBTcl_DStringStartSublist\fR(\fIdsPtr\fR)
.sp
\fBTcl_DStringEndSublist\fR(\fIdsPtr\fR)
.sp
int
\fBTcl_DStringLength\fR(\fIdsPtr\fR)
.sp
char *
\fBTcl_DStringValue\fR(\fIdsPtr\fR)
.sp
\fBTcl_DStringSetLength\fR(\fIdsPtr, newLength\fR)
.sp
\fBTcl_DStringTrunc\fR(\fIdsPtr, newLength\fR)
.sp
\fBTcl_DStringFree\fR(\fIdsPtr\fR)
.sp
\fBTcl_DStringResult\fR(\fIinterp, dsPtr\fR)
.sp
\fBTcl_DStringGetResult\fR(\fIinterp, dsPtr\fR)
.SH ARGUMENTS
.AS Tcl_DString newLength
.AP Tcl_DString *dsPtr in/out
Pointer to structure that is used to manage a dynamic string.
.AP char *string in
Pointer to characters to add to dynamic string.
.AP int length in
Number of characters from string to add to dynamic string.  If -1,
add all characters up to null terminating character.
.AP int newLength in
New length for dynamic string, not including null terminating
character.
.AP Tcl_Interp *interp in/out
Interpreter whose result is to be set from or moved to the
dynamic string.
.BE

.SH DESCRIPTION
.PP
Dynamic strings provide a mechanism for building up arbitrarily long
strings by gradually appending information.  If the dynamic string is
short then there will be no memory allocation overhead;  as the string
gets larger, additional space will be allocated as needed.
.PP
\fBTcl_DStringInit\fR initializes a dynamic string to zero length.
The Tcl_DString structure must have been allocated by the caller.
No assumptions are made about the current state of the structure;
anything already in it is discarded.
If the structure has been used previously, \fBTcl_DStringFree\fR should
be called first to free up any memory allocated for the old
string.
.PP
\fBTcl_DStringAppend\fR adds new information to a dynamic string,
allocating more memory for the string if needed.
If \fIlength\fR is less than zero then everything in \fIstring\fR
is appended to the dynamic string;  otherwise \fIlength\fR
specifies the number of bytes to append.
\fBTcl_DStringAppend\fR returns a pointer to the characters of
the new string.  The string can also be retrieved from the
\fIstring\fR field of the Tcl_DString structure.
.PP
\fBTcl_DStringAppendElement\fR is similar to \fBTcl_DStringAppend\fR
except that it doesn't take a \fIlength\fR argument (it appends
all of \fIstring\fR) and it converts the string to a proper list element
before appending.
\fBTcl_DStringAppendElement\fR adds a separator space before the
new list element unless the new list element is the first in a
list or sub-list (i.e. either the current string is empty, or it
contains the single character ``{'', or the last two characters of
the current string are `` {'').
\fBTcl_DStringAppendElement\fR returns a pointer to the
characters of the new string.
.PP
\fBTcl_DStringStartSublist\fR and \fBTcl_DStringEndSublist\fR can be
used to create nested lists.
To append a list element that is itself a sublist, first
call \fBTcl_DStringStartSublist\fR, then call \fBTcl_DStringAppendElement\fR
for each of the elements in the sublist, then call
\fBTcl_DStringEndSublist\fR to end the sublist.
\fBTcl_DStringStartSublist\fR appends a space character if needed,
followed by an open brace;  \fBTcl_DStringEndSublist\fR appends
a close brace.
Lists can be nested to any depth.
.PP
\fBTcl_DStringLength\fR is a macro that returns the current length
of a dynamic string (not including the terminating null character).
\fBTcl_DStringValue\fR is a  macro that returns a pointer to the
current contents of a dynamic string.
.PP
.PP
\fBTcl_DStringSetLength\fR changes the length of a dynamic string.
If \fInewLength\fR is less than the string's current length, then
the string is truncated.
If \fInewLength\fR is greater than the string's current length,
then the string will become longer and new space will be allocated
for the string if needed.
However, \fBTcl_DStringSetLength\fR will not initialize the new
space except to provide a terminating null character;  it is up to the
caller to fill in the new space.
\fBTcl_DStringSetLength\fR does not free up the string's storage space
even if the string is truncated to zero length, so \fBTcl_DStringFree\fR
will still need to be called.
.PP
\fBTcl_DStringTrunc\fR changes the length of a dynamic string.
This procedure is now deprecated.  \fBTcl_DStringSetLength\fR  should
be used instead.
.PP
\fBTcl_DStringFree\fR should be called when you're finished using
the string.  It frees up any memory that was allocated for the string
and reinitializes the string's value to an empty string.
.PP
\fBTcl_DStringResult\fR sets the result of \fIinterp\fR to the value of
the dynamic string given by \fIdsPtr\fR.  It does this by moving
a pointer from \fIdsPtr\fR to the interpreter's result.
This saves the cost of allocating new memory and copying the string.
\fBTcl_DStringResult\fR also reinitializes the dynamic string to
an empty string.
.PP
\fBTcl_DStringGetResult\fR does the opposite of \fBTcl_DStringResult\fR.
It sets the value of \fIdsPtr\fR to the result of \fIinterp\fR and
it clears \fIinterp\fR's result.
If possible it does this by moving a pointer rather than by copying
the string.

.SH KEYWORDS
append, dynamic string, free, result


ChangeLog
2004-06-14 20:28  Dossy Shiobara <dossy@panoptic.com>

	* nsd/connio.c: If nsend == 0, we would never call Ns_WriteConn to
	  flush the queued headers.  This could happen when we're sending
	  zero bytes of data as a response.  This fixes bug #971016 in the
	  case where fastpath.cache=false and fastpath.mmap=false and a
	  zero byte file is requested.

2004-06-14 20:25  Dossy Shiobara <dossy@panoptic.com>

	* nsd/return.c: unnecessary test for data != NULL actually caused
	  part of bug #971016 when fastpath.cache=false and
	  fastpath.mmap=true and the file requested is zero bytes, mmap()
	  returns 0 which gets passed along as data == NULL, causing
	  ReturnCharData() to not flush the queued headers.  removing the
	  if is safe as Ns_WriteConn will simply flush any queued data

2004-06-08  Rob Crittenden <rcrittenden0569@aol.com>

	* Makefile: remove Makefile.module on distclean.
	* include/ns.h, nsd/log.c: Allow users to override logging functions.
	* nsd/str.c: Fix crash bug in Ns_Trim* when trimming NULL strings.

2004-06-08  Rob Crittenden <rcrittenden0569@aol.com>

	* TAG aolserver_v40_r5

2004-03-09 13:35  Rob Crittenden <rcrittenden0569@aol.com>

	* nsd/tclhttp.c: The last header was being missed because the pointer
	  was being advanced past the \r\n.

2004-02-26 14:37  Dossy Shiobara <dossy@panoptic.com>

	* nsd/unix.c: uid_t is supposed to be unsigned, but apparently
	  isn't on OSX, apparently.

2004-02-25 19:56  Dossy Shiobara <dossy@panoptic.com>

	* nscgi/nscgi.c: Ns_ConnSetRequiredHeaders now requires length < 0
	  if you don't want a Content-Length header

2004-02-25 19:55  Dossy Shiobara <dossy@panoptic.com>

	* nsd/return.c: fix so that it's still possible to formulate a
	  response with no Content-Length header (such as for nscgi, or
	  HTTP streaming) where you don't know the Content-Length header in
	  advance

2004-02-24 07:15  Dossy Shiobara <dossy@panoptic.com>

	* nsd/: nsmain.c, nsd.h, driver.c: fix race condition where
	  DriverThreads are spawned but before they can Ns_SockListenEx,
	  the main thread does its NsClosePreBound and closes the prebound
	  sockets, making them unavailable to the DriverThreads if they
	  require root priv to listen on

2004-02-18 01:30  Dossy Shiobara <dossy@panoptic.com>

	* nsperm/nsperm.c: fixed bug #899364, restoring old nsperm
	  functionality as it was in 3.x.

2004-02-15  Jim Davidson <jgdavidson@aol.com>

	* nsd/conn.c:
	* include/ns.h:
	* nsd/return.c: Added new Ns_ConnGet/SetKeepAliveFlag APIs to
	allow custom request procedures to enable keep-alive directly.

	* nsd/driver.c: Minor updates to separate conn & driver structs.

2004-01-01  Scott S. Goodwin <scott@scottg.net>

        * nsd/adpparse.c: fixed bug #868600, where tag attribute values
	encapsulated within single quotes weren't being parsed properly.

2003-12-27  Scott S. Goodwin <scott@scottg.net>

        * nsd/urlopen.c: fixed bug #844809.

        * sample-config.tcl: fixed bug #805761.

2003-12-24  Scott S. Goodwin <scott@scottg.net>

	* include/nsthread.h: Bumped up NS_THREAD_NAMESIZE from 32 to 64.
	Mutex names were being truncated, making it impossible to distinguish
	mutexes with longer names when looking at stats.

        * nscgi/nscgi.c: Changed HTTPS env variable to be set to "on" or "off"
	instead of "1" or "0". There doesn't appear to be any standard for
	this, but Apache uses the on/off string in the environment.

2003-12-16  Scott S. Goodwin <scott@scottg.net>

        * nscgi/nscgi.c: Fixed issue with HTTPS environment variable not being
        set, and port not being set to 443 if connection is SSL'd. Manually
        applied these changes from a patch supplied by Vinod Kurup.

2003-11-16  Jim Davidson <jgdavidson@aol.com>

	* include/ns.h: Update version ids, added NS_SOCK_DROP and
	NS_TCL_TRACE constants and pre-queue function definitions.

	* nsd/nsd.h:  Added new "times" struct to Conn for more
	fine-grained timing.  Need to provide options to enable
	these time stamps plus ways to get the data.

	* nsd/driver.c:  Restructured in a few ways:

	1. Each loaded driver now has it's own thread.  This should
	allow some more concurrency (if necessary) and generally
	made the code more manageable.

	2. New pre-queue I/O callback API.  The API will be documented
	with an example HTTP module later.

	3. Conn allocation is handled by the driver threads and
	request buffers have been moved to the Conn from the Sock.

	4. Reads for synchrounous drivers (i.e., SSL) are handled
	by a pool of reader threads instead of by the conn threads.
	The number of threads is set with a "maxreaders" config
	option.  This change should enable both Host: header based
	virtual servers and pre-queue callbacks for SSL connections.

	5. New "maxsocks" config for drivers is now used to manage
	the concurrency limit of the server.  Conn's are now queued
	without limit into the connection pools and the previous
	"maxconns" virtual-server config is ignored.

	* nsd/queue.c: Cleaned up "connsperthread" code and updated
	NsQueueConn to support new driver code.

	* nsd/conn.c:
	* nsd/connio.c:
	* nsd/form.c:
	* nsd/nsd.h:
	* nsd/queue.c:
	* nsd/return.c: Moved connection request buffers from Sock
	to Conn structure.

	* nsd/server.c: Removed startup Conn pre-allocation as Conns
	are now allocated dynamically by the driver.  Also, removed
	the "maxconns" parameter which is no longer used.

	* nssock/nssock.c: Added new "async" option (default on)
	mostly to test new reader thread capabilities of the core
	driver code.

	* nsd/tclinit.c: Added a new generalized Tcl interp tracing
	facility and cleaned up some previous init code.  This
	should make it easier to support standard Tcl extensions
	and will be documented with an example module later.

	* nsd/nsmain.c: Fixed order of driver shutdown wait.

	* nsd/init.c: Fixed order of init calls.

	* nsd/binder.c:
	* nsd/dns.c: Named unnamed mutexes.

	* nsd/tclset.c:
	* nsd/tclsock.c: Removed unused functions.


nsd/dstring.c
/*
 * The contents of this file are subject to the AOLserver Public License
 * Version 1.1 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 * http://aolserver.com/.
 *
 * Software distributed under the License is distributed on an "AS IS"
 * basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See
 * the License for the specific language governing rights and limitations
 * under the License.
 *
 * The Original Code is AOLserver Code and related documentation
 * distributed by AOL.
 * 
 * The Initial Developer of the Original Code is America Online,
 * Inc. Portions created by AOL are Copyright (C) 1999 America Online,
 * Inc. All Rights Reserved.
 *
 * Alternatively, the contents of this file may be used under the terms
 * of the GNU General Public License (the "GPL"), in which case the
 * provisions of GPL are applicable instead of those above.  If you wish
 * to allow use of your version of this file only under the terms of the
 * GPL and not to allow others to use your version of this file under the
 * License, indicate your decision by deleting the provisions above and
 * replace them with the notice and other provisions required by the GPL.
 * If you do not delete the provisions above, a recipient may use your
 * version of this file under either the License or the GPL.
 */


/* 
 * dstring.c --
 *
 *	Ns_DString routines.  Ns_DString's are not compatible 
 *	with Tcl_DString's.
 */

static const char *RCSID = "@(#) $Header: /Users/dossy/Desktop/cvs/aolserver/nsd/dstring.c,v 1.7 2000/12/14 21:57:28 kriston Exp $, compiled: " __DATE__ " " __TIME__;

#include "nsd.h"

/*
 * The following structure is used to maintain the per-thread
 * DString cache.
 */
 
typedef struct {
    Ns_DString *firstPtr;
    int ncached;
} Stack;

static Ns_Tls tls;  	    	    /* Cache TLS. */
static Ns_Callback FlushDStrings;   /* Cache TLS cleannup. */
static void GrowDString(Ns_DString *dsPtr, int length);


/*
 *----------------------------------------------------------------------
 * Ns_DStringInit --
 *
 *      Initialize a Ns_DString object.
 *
 * Results:
 *      None.
 *
 * Side effects:
 *      None.
 *
 *----------------------------------------------------------------------
 */

void
Ns_DStringInit(Ns_DString *dsPtr)
{
    dsPtr->string = dsPtr->staticSpace;
    dsPtr->length = 0;
    dsPtr->spaceAvl = NS_DSTRING_STATIC_SIZE;
    dsPtr->staticSpace[0] = 0;
    dsPtr->addr = NULL;
}


/*
 *----------------------------------------------------------------------
 *
 * Ns_DStringVarAppend --
 *
 *	Append a variable number of string arguments to a dstring.
 *
 * Results:
 *	Pointer to current dstring value.
 *
 * Side effects:
 *	None.
 *
 *----------------------------------------------------------------------
 */

char *
Ns_DStringVarAppend(Ns_DString *dsPtr, ...)
{
    register char   *s;
    va_list         ap;

    va_start(ap, dsPtr);
    while ((s = va_arg(ap, char *)) != NULL) {
	Ns_DStringAppend(dsPtr, s);
    }
    va_end(ap);

    return dsPtr->string;
}


/*
 *----------------------------------------------------------------------
 *
 * Ns_DStringAppendElement --
 *
 *	Append a list element to the current value of a dynamic string.
 *
 * Results:
 *	The return value is a pointer to the dynamic string's new value.
 *
 * Side effects:
 *	String is reformatted as a list element and added to the current
 *	value of the string.  Memory gets reallocated if needed to
 *	accomodate the string's new size.
 *
 *----------------------------------------------------------------------
 */

char *
Ns_DStringAppendElement(Ns_DString *dsPtr, char *string)
{
    int newSize, flags;
    char *dst;
    extern int TclNeedSpace(char *, char *);

    newSize = Tcl_ScanElement(string, &flags) + dsPtr->length + 1;

    /*
     * Grow the string if necessary.
     */

    if (newSize >= dsPtr->spaceAvl) {
	GrowDString(dsPtr, newSize * 2);
    }

    /*
     * Convert the new string to a list element and copy it into the
     * buffer at the end, with a space, if needed.
     */

    dst = dsPtr->string + dsPtr->length;
    if (TclNeedSpace(dsPtr->string, dst)) {
	*dst = ' ';
	dst++;
	dsPtr->length++;
    }
    dsPtr->length += Tcl_ConvertElement(string, dst, flags);
    return dsPtr->string;
}


/*
 *----------------------------------------------------------------------
 *
 * Ns_DStringExport --
 *
 *	Return a copy of the string value on the heap.
 *	Ns_DString is left in an initialized state.
 *
 * Results:
 *	Pointer to ns_malloc'ed string which must be eventually freed.
 *
 * Side effects:
 *	None.
 *
 *
 *----------------------------------------------------------------------
 */

char *
Ns_DStringExport(Ns_DString *dsPtr)
{
    char *s;

    if (dsPtr->string != dsPtr->staticSpace) {
	s = dsPtr->string;
	dsPtr->string = dsPtr->staticSpace;
    } else {
	s = ns_malloc(dsPtr->length+1);
	memcpy(s, dsPtr->string, dsPtr->length+1);  
    }
    Ns_DStringFree(dsPtr);

    return s;
}


/*
 *----------------------------------------------------------------------
 * Ns_DStringPrintf --
 *
 *      Append a sequence of values using a format string
 *
 * Results:
 *	Pointer to the current string value.
 *
 * Side effects:
 *      None.
 *
 *----------------------------------------------------------------------
 */

char *
Ns_DStringPrintf(Ns_DString *dsPtr, char *fmt,...)
{
    char            buf[DSTRINGPRINTF_LIMIT];
    va_list         ap;

    va_start(ap, fmt);
    vsnprintf(buf, DSTRINGPRINTF_LIMIT, fmt, ap);
    va_end(ap);

    return Ns_DStringAppend(dsPtr, buf);
}


/*
 *----------------------------------------------------------------------
 * Ns_DStringFree --
 *
 *	Reset to its initialized state and deallocate any allocated
 *	memory.
 *
 * Results:
 *      None.
 *
 * Side effects:
 *      None.
 *
 *----------------------------------------------------------------------
 */

void
Ns_DStringFree(Ns_DString *dsPtr)
{
    if (dsPtr->string != dsPtr->staticSpace) {
	ns_free(dsPtr->string);
    }
    dsPtr->string = dsPtr->staticSpace;
    dsPtr->length = 0;
    dsPtr->spaceAvl = NS_DSTRING_STATIC_SIZE;
    dsPtr->staticSpace[0] = 0;
}


/*
 *----------------------------------------------------------------------
 *
 * Ns_DStringSetLength --
 *
 *	Change the length of a dynamic string.  This can cause the
 *	string to either grow or shrink, depending on the value of
 *	length.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	The length of dsPtr is changed to length and a null byte is
 *	stored at that position in the string.  If length is larger
 *	than the space allocated for dsPtr, then a panic occurs.
 *
 *----------------------------------------------------------------------
 */

void
Ns_DStringSetLength(Ns_DString *dsPtr, int length)
{
    if (length < 0) {
	length = 0;
    }
    if (length >= dsPtr->spaceAvl) {
	GrowDString(dsPtr, length+1);
    }
    dsPtr->length = length;
    dsPtr->string[length] = 0;
}


/*
 *----------------------------------------------------------------------
 * Ns_DStringTrunc --
 *
 *      Truncate the value to a null string.
 *
 * Results:
 *      None.
 *
 * Side effects:
 *      None.
 *
 *----------------------------------------------------------------------
 */

void
Ns_DStringTrunc(Ns_DString *dsPtr, int length)
{
    Ns_DStringSetLength(dsPtr, length);
}


/*
 *----------------------------------------------------------------------
 * Ns_DStringNAppend --
 *
 *      Append a byte copy of the first length bytes of string.
 *
 * Results:
 *	Pointer to the current string value.
 *
 * Side effects:
 *      None.
 *
 *----------------------------------------------------------------------
 */

char *
Ns_DStringNAppend(Ns_DString *dsPtr, char *string, int length)
{
    int newSize;

    if (length < 0) {
	length = strlen(string);
    }
    newSize = length + dsPtr->length;

    /*
     * Grow the string if necessary.
     */

    if (newSize >= dsPtr->spaceAvl) {
	GrowDString(dsPtr, newSize * 2);
    }

    /*
     * Copy the new string into the buffer at the end of the old one.
     */

    memcpy(dsPtr->string + dsPtr->length, string, length);
    dsPtr->length += length;
    dsPtr->string[dsPtr->length] = '\0';

    return dsPtr->string;
}


/*
 *----------------------------------------------------------------------
 * Ns_DStringAppendArg --
 *
 *      Append a string including its terminating null byte.
 *
 * Results:
 *	Pointer to the current string value.
 *
 * Side effects:
 *      None.
 *
 *----------------------------------------------------------------------
 */

char *
Ns_DStringAppendArg(Ns_DString *dsPtr, char *string)
{
    return Ns_DStringNAppend(dsPtr, string, strlen(string) + 1);
}


/*
 *----------------------------------------------------------------------
 * Ns_DStringAppend --
 *
 *      Append a string.
 *
 * Results:
 *	Pointer to the current string value.
 *
 * Side effects:
 *      None.
 *
 *----------------------------------------------------------------------
 */

#undef Ns_DStringAppend

char *
Ns_DStringAppend(Ns_DString *dsPtr, char *string)
{
    return Ns_DStringNAppend(dsPtr, string, -1);
}


/*
 *----------------------------------------------------------------------
 * Ns_DStringLength --
 *
 *      Return the length of a string.
 *
 * Results:
 *      The length of a string.
 *
 * Side effects:
 *      None.
 *
 *----------------------------------------------------------------------
 */

#undef Ns_DStringLength

int
Ns_DStringLength(Ns_DString *dsPtr)
{
    return dsPtr->length;
}


/*
 *----------------------------------------------------------------------
 * Ns_DStringValue --
 *
 *	Return a pointer to the current string value.
 *
 * Results:
 *	Pointer to the current string value.
 *
 * Side effects:
 *      None.
 *
 *----------------------------------------------------------------------
 */

#undef Ns_DStringValue

char *
Ns_DStringValue(Ns_DString *dsPtr)
{
    return dsPtr->string;
}


/*
 *----------------------------------------------------------------------
 * Ns_DStringPop --
 *
 *	Pop a DString from the per-thread cache, allocating one
 *  	if necessary.
 *
 * Results:
 *	Pointer to initialized DString.
 *
 * Side effects:
 *      Will initialize TLS and config parameters on first use.
 *
 *----------------------------------------------------------------------
 */

Ns_DString *
Ns_DStringPop(void)
{
    Stack *sPtr;
    Ns_DString *dsPtr;

    if (tls == NULL) {
	Ns_MasterLock();
	if (tls == NULL) {
	    Ns_TlsAlloc(&tls, FlushDStrings);
	}
	Ns_MasterUnlock();
    }
    sPtr = Ns_TlsGet(&tls);
    if (sPtr == NULL) {
	sPtr = ns_calloc(1, sizeof(Stack));
	Ns_TlsSet(&tls, sPtr);
    }
    if (sPtr->firstPtr == NULL) {
	dsPtr = ns_malloc(sizeof(Ns_DString));
	Ns_DStringInit(dsPtr);
    } else {
	dsPtr = sPtr->firstPtr;
	sPtr->firstPtr = *((Ns_DString **) dsPtr->staticSpace);
    	dsPtr->staticSpace[0] = 0;
	--sPtr->ncached;
    }
    return dsPtr;
}


/*
 *----------------------------------------------------------------------
 * Ns_DStringPush --
 *
 *	Push a DString back on the per-thread cache.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *      DString may be free'ed if the max entries and/or size has been
 *  	exceeded.
 *
 *----------------------------------------------------------------------
 */

void
Ns_DStringPush(Ns_DString *dsPtr)
{
    Stack *sPtr = Ns_TlsGet(&tls);

    if (sPtr->ncached >= nsconf.dstring.maxentries) {
	Ns_DStringFree(dsPtr);
	ns_free(dsPtr);
    } else {
	if (dsPtr->spaceAvl > nsconf.dstring.maxsize) {
	    Ns_DStringFree(dsPtr);
	} else {
    	    Ns_DStringTrunc(dsPtr, 0);
	}
	*((Ns_DString **) dsPtr->staticSpace) = sPtr->firstPtr;
	sPtr->firstPtr = dsPtr;
	++sPtr->ncached;
    }
}


/*
 *----------------------------------------------------------------------
 * FlushDStrings --
 *
 *	TLS callback to flush the DString cache.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *  	None.
 *
 *----------------------------------------------------------------------
 */

static void
FlushDStrings(void *arg)
{
    Stack *sPtr = arg;
    Ns_DString *dsPtr;

    while ((dsPtr = sPtr->firstPtr) != NULL) {
	sPtr->firstPtr = *((Ns_DString **) dsPtr->staticSpace);
	Ns_DStringFree(dsPtr);
	ns_free(dsPtr);
    }
    ns_free(sPtr);
}


/*
 *----------------------------------------------------------------------
 * GrowDString --
 *
 *	Increase the space available in a dstring.  Note that
 *	memcpy() is used instead of strcpy() because the string
 *	may have embedded nulls.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *  	New memory will be allocated.
 *
 *----------------------------------------------------------------------
 */

static void
GrowDString(Ns_DString *dsPtr, int length)
{
    char *newString;

    if (dsPtr->string != dsPtr->staticSpace) {
	newString = ns_realloc(dsPtr->string, (size_t) length);
    } else {
	newString = ns_malloc(length);
	memcpy(newString, dsPtr->string, (size_t) dsPtr->length);
    }
    dsPtr->string = newString;
    dsPtr->spaceAvl = length;
}
